crates/cosmos/cosmos-client-components/src/impls/queries/connection_end.rs

```diff
@@ -0,0 +1,53 @@
+use cgp_core::HasErrorType;
+use hermes_relayer_components::chain::traits::queries::connection_end::ConnectionEndQuerier;
+use hermes_relayer_components::chain::traits::types::connection::HasConnectionEndType;
+use hermes_relayer_components::chain::traits::types::ibc::HasIbcChainTypes;
+use ibc_relayer::chain::handle::ChainHandle;
+use ibc_relayer::chain::requests::IncludeProof;
+use ibc_relayer::chain::requests::QueryConnectionRequest;
+use ibc_relayer::chain::requests::QueryHeight;
+use ibc_relayer_types::core::ics03_connection::connection::ConnectionEnd;
+use ibc_relayer_types::core::ics24_host::identifier::ConnectionId;
+use ibc_relayer_types::Height;
+
+use crate::traits::chain_handle::HasBlockingChainHandle;
+
+pub struct QueryCosmosConnectionEndFromChainHandle;
+
+impl<Chain, Counterparty> ConnectionEndQuerier<Chain, Counterparty>
+    for QueryCosmosConnectionEndFromChainHandle
+where
+    Chain: HasConnectionEndType<Counterparty, ConnectionEnd = ConnectionEnd>
+        + HasIbcChainTypes<Counterparty, Height = Height, ConnectionId = ConnectionId>
+        + HasBlockingChainHandle
+        + HasErrorType,
+{
+    async fn query_connection_end(
+        chain: &Chain,
+        connection_id: &Chain::ConnectionId,
+        height: Option<&Chain::Height>,
```

Comment:
Is this the simplest way to get the connection end at the latest height without incurring another redundant query at the call site when the user is only interested in the latest height?

I would vote for either keeping the optional height parameter or define our own `enum QueryHeight<Height> { Latest, Specific(Height) }` type and take a `QueryHeight<Chain::Height>` as an argument.

Reply:
The chain height can easily be cached every few hundred milliseconds, so that the query do not become costly.

If we really want to make it optional, perhaps we can add an abstract `HeightParam` as follows:

```rust
pub trait HasHeightParamType: HasHeightType {
    type HeightParam: Async + From<Self::Height>;
}
```

We would also add `WithLatestHeight` variants of the trait, and allow implementations to choose different strategies to handle the latest height parameter. For example:

```
impl<Chain> ClientStateWithLatestHeightQuerier<Chain>
    for QueryLatestClientStateWithDefaultHeightParam
where
    Chain::HeightParam: Default,
```

and

```
impl<Chain> ClientStateWithLatestHeightQuerier<Chain>
    for QueryLatestHeightAndClientState
where
    Chain: CanQueryChainHeight,
```

Anyway, it is not too important for the current case of querying connection end, as it is only used once in the CLI. Let's keep this for now, and continue our discussion somewhere else.

Reply:
Opened https://github.com/informalsystems/hermes-sdk/issues/172